kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/AdminManager.java

```diff
@@ -439,10 +445,30 @@ private void handleUpdatePartitionsAsync(String topic,
     public void setBrokers(Map<String, Set<Node>> newBrokers) {
         brokersCacheLock.writeLock().lock();
         try {
+            setControllerId(newBrokers);
             this.brokersCache = newBrokers;
         } finally {
             brokersCacheLock.writeLock().unlock();
         }
     }
 
+    // only set when setBrokers
+    private void setControllerId(Map<String, Set<Node>> newBrokers) {
+        Map<String, Integer> newControllerId = Maps.newHashMap();
+        newBrokers.forEach((listenerName, brokers) -> {
+            if (brokers.size() == 0) {
+                newControllerId.put(listenerName, MetadataResponse.NO_CONTROLLER_ID);
+            } else {
+                List<Node> nodes = Lists.newArrayList(brokers);
+                newControllerId.put(listenerName,
+                        nodes.size() > 1 ? nodes.get(random.nextInt(brokers.size())).id() : nodes.get(0).id());
+            }
+        });
+        this.controllerId = newControllerId;
+    }
+
+    // always get the controllerId directly from the cache
+    public int getControllerId(String listenerName) {
+        return controllerId.getOrDefault(listenerName, MetadataResponse.NO_CONTROLLER_ID);
```

Comment:
Could you ensure `getControllerId` and `setControllerId` are called in the same thread? If not, it would be not thread-safe without the lock. The `volatile` keyword can only ensure the `controllerId` is the latest reference, but you can't ensure a `null` value is read.

Reply:
Outdated `controllerId` is allowed, this will not cause any problems, it is better for the client to briefly retry the metadata request than the server to use lock synchronization, otherwise we may see all the request handle threads are stuck getting the `controllerId` , and the `KopEventThread` is updating the `controllerId` now.

The possible consistency issue here is the visibility order of `controllerId` and `brokerCache`, if you are very strong about this, we can define a `Metadata` or `MetadataCache` class to wrap `controllerId` and `brokerCache` for synchronization purposes.

Reply:
| Time | Operation                                                    | `controllerId` |
| ---- | ------------------------------------------------------------ | -------------- |
| 1    | setControllerId({"A": 1})                                    | {"A": 1}       |
| 2    | getControllerId("B") read `controllerId`: {"A": 1}           | {"A": 1}       |
| 3    | setControllerId({"B": 2})                                    | {"B": 2}       |
| 4    | getControllerId("B") **getOrDefault: operation on an outdated reference** |                |

I'm not sure whether the operation on an outdated reference is valid.

Reply:
Get controllerId in kafka:
https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/server/metadata/ZkMetadataCache.scala#L297